File: src/app/flutter/convention_list/lib/models/search_params.dart
import 'package:convention_list/models/position.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'search_params.freezed.dart';
part 'search_params.g.dart';

@freezed
class SearchParams with _$SearchParams {
  const factory SearchParams({
    @Default(OrderBy.distance) OrderBy orderBy,
    String? search,
    Position? position,
  }) = _SearchParams;

  const SearchParams._();

  String toQueryString() {
    String queryString = '&orderBy=${orderBy.name}';
    if (search != null && search!.isNotEmpty) {
      queryString += '&search=$search';
    }

    if (position != null) {
      queryString += '&lat=${position!.latitude}&lon=${position!.longitude}';
    }
    return queryString;
  }

  factory SearchParams.fromJson(Map<String, dynamic> json) => _$SearchParamsFromJson(json);
}

enum OrderBy {
  distance,
  startDate,
}

File: src/reactclient/src/SearchParams.js
export default class SearchParams {
  constructor({
    category,
    search,
    startRange,
    endRange,
    lat,
    lon,
    orderBy,
    page,
  }) {
    this.category = category;
    this.search = search;
    this.startRange = startRange;
    this.endRange = endRange;
    this.lat = lat;
    this.lon = lon;
    this.orderBy = orderBy;
    this.page = page;
  }

  toQueryString() {
    const queryStringArray = [];
    if (this.category) {
      queryStringArray.push(`category=${this.category}`);
    }

    if (this.search) {
      queryStringArray.push(`search=${this.search}`);
    }

    if (this.startRange) {
      queryStringArray.push(`startRange=${this.startRange}`);
    }

    if (this.endRange) {
      queryStringArray.push(`endRange=${this.endRange}`);
    }

    if (this.lat) {
      queryStringArray.push(`lat=${this.lat}`);
    }

    if (this.lon) {
      queryStringArray.push(`lon=${this.lon}`);
    }

    if (this.orderBy) {
      queryStringArray.push(`orderBy=${this.orderBy}`);
    }

    if (this.page) {
      queryStringArray.push(`page=${this.page}`);
    }

    if (queryStringArray.length === 0) {
      return "";
    } else {
      return `?${queryStringArray.join("&")}`;
    }
  }
}
